### ISSUE TITLE
Setting crashing due to battery stats

### ISSUE THREAD

**Issue by yihuaf** [anbox/anbox#767]
5. Run the anbox logs collection utility and attach the tar file.
`sudo /snap/bin/anbox.collect-bug-info`
[anbox-system-diagnostics-2018-06-21.tar.xz.zip](https://github.com/anbox/anbox/files/2124785/anbox-system-diagnostics-2018-06-21.tar.xz.zip)

6. ** Please paste the result of `anbox system-info` below:**
```
version: 4
snap-revision: 108
os:
  name: Ubuntu
  version: 16.04.4 LTS (Xenial Xerus)
  snap-based: true
kernel:
  version: Linux version 4.13.0-45-generic (buildd@lgw01-amd64-011) (gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.9)) #50~16.04.1-Ubuntu SMP Wed May 30 11:18:27 UTC 2018
  binder: true
  ashmem: true
graphics:
  egl:
    vendor: Mesa Project
    version: 1.4 (DRI2)
    extensions:
      - EGL_CHROMIUM_sync_control
      - EGL_EXT_image_dma_buf_import
      - EGL_KHR_cl_event2
      - EGL_KHR_config_attribs
      - EGL_KHR_create_context
      - EGL_KHR_create_context_no_error
      - EGL_KHR_fence_sync
      - EGL_KHR_get_all_proc_addresses
      - EGL_KHR_gl_colorspace
      - EGL_KHR_gl_renderbuffer_image
      - EGL_KHR_gl_texture_2D_image
      - EGL_KHR_gl_texture_3D_image
      - EGL_KHR_gl_texture_cubemap_image
      - EGL_KHR_image
      - EGL_KHR_image_base
      - EGL_KHR_image_pixmap
      - EGL_KHR_no_config_context
      - EGL_KHR_reusable_sync
      - EGL_KHR_surfaceless_context
      - EGL_KHR_wait_sync
      - EGL_MESA_configless_context
      - EGL_MESA_drm_image
      - EGL_MESA_image_dma_buf_export
      - EGL_NOK_swap_region
      - EGL_NOK_texture_from_pixmap
      - EGL_NV_post_sub_buffer
      - EGL_WL_bind_wayland_display
  gles2:
    vendor: nouveau
    vendor: OpenGL ES 3.1 Mesa 17.2.8
    extensions:
      - GL_EXT_blend_minmax
      - GL_EXT_multi_draw_arrays
      - GL_EXT_texture_filter_anisotropic
      - GL_EXT_texture_format_BGRA8888
      - GL_OES_compressed_ETC1_RGB8_texture
      - GL_OES_depth24
      - GL_OES_element_index_uint
      - GL_OES_fbo_render_mipmap
      - GL_OES_mapbuffer
      - GL_OES_rgb8_rgba8
      - GL_OES_standard_derivatives
      - GL_OES_stencil8
      - GL_OES_texture_3D
      - GL_OES_texture_float
      - GL_OES_texture_float_linear
      - GL_OES_texture_half_float
      - GL_OES_texture_half_float_linear
      - GL_OES_texture_npot
      - GL_OES_vertex_half_float
      - GL_EXT_texture_sRGB_decode
      - GL_OES_EGL_image
      - GL_OES_depth_texture
      - GL_OES_packed_depth_stencil
      - GL_EXT_texture_type_2_10_10_10_REV
      - GL_OES_get_program_binary
      - GL_APPLE_texture_max_level
      - GL_EXT_discard_framebuffer
      - GL_EXT_read_format_bgra
      - GL_EXT_frag_depth
      - GL_NV_fbo_color_attachments
      - GL_OES_EGL_image_external
      - GL_OES_EGL_sync
      - GL_OES_vertex_array_object
      - GL_OES_viewport_array
      - GL_EXT_robustness
      - GL_EXT_texture_rg
      - GL_EXT_unpack_subimage
      - GL_NV_draw_buffers
      - GL_NV_read_buffer
      - GL_NV_read_depth
      - GL_NV_read_depth_stencil
      - GL_NV_read_stencil
      - GL_EXT_draw_buffers
      - GL_EXT_map_buffer_range
      - GL_KHR_debug
      - GL_KHR_robustness
      - GL_OES_depth_texture_cube_map
      - GL_OES_surfaceless_context
      - GL_EXT_color_buffer_float
      - GL_EXT_separate_shader_objects
      - GL_EXT_shader_integer_mix
      - GL_EXT_tessellation_point_size
      - GL_EXT_tessellation_shader
      - GL_EXT_base_instance
      - GL_EXT_compressed_ETC1_RGB8_sub_texture
      - GL_EXT_copy_image
      - GL_EXT_draw_buffers_indexed
      - GL_EXT_draw_elements_base_vertex
      - GL_EXT_gpu_shader5
      - GL_EXT_polygon_offset_clamp
      - GL_EXT_primitive_bounding_box
      - GL_EXT_shader_io_blocks
      - GL_EXT_texture_border_clamp
      - GL_EXT_texture_buffer
      - GL_EXT_texture_cube_map_array
      - GL_KHR_blend_equation_advanced
      - GL_KHR_context_flush_control
      - GL_KHR_robust_buffer_access_behavior
      - GL_NV_image_formats
      - GL_OES_copy_image
      - GL_OES_draw_buffers_indexed
      - GL_OES_draw_elements_base_vertex
      - GL_OES_gpu_shader5
      - GL_OES_primitive_bounding_box
      - GL_OES_sample_shading
      - GL_OES_sample_variables
      - GL_OES_shader_io_blocks
      - GL_OES_shader_multisample_interpolation
      - GL_OES_tessellation_point_size
      - GL_OES_tessellation_shader
      - GL_OES_texture_border_clamp
      - GL_OES_texture_buffer
      - GL_OES_texture_cube_map_array
      - GL_OES_texture_stencil8
      - GL_OES_texture_storage_multisample_2d_array
      - GL_EXT_blend_func_extended
      - GL_EXT_buffer_storage
      - GL_EXT_geometry_point_size
      - GL_EXT_geometry_shader
      - GL_KHR_no_error
      - GL_OES_geometry_point_size
      - GL_OES_geometry_shader
      - GL_OES_shader_image_atomic
      - GL_EXT_clip_cull_distance
      - GL_EXT_window_rectangles
      - GL_MESA_shader_integer_functions
```

**Please describe your problem:**
When clicking on battery tab under Settings or click into app info, the setting app crashes and the logcat shows `BatteryStatsHelper` is crashing. The stack trace of the exception is attached in additional info. It looks like a memory file read for android stats stream returns `EINVAL` and caused the crash. Not sure if this is related to `ashmem` or `binde`r. I have these two kernel module installed from the ppa following the anbox install instruction. I attached the kernel mod info for your information in the section below.

The setting->app and then click on any tab also reproduce this issue with the same stack trace, crashing inside the `BatteryStatsHelper`.

**What were you expecting?:**

I should be able to click into setting battery and into app to change permission of the app.

**Additional info:**

```
06-21 17:30:39.322   330   330 W BatteryStatsHelper: Unable to read statistics stream
06-21 17:30:39.322   330   330 W BatteryStatsHelper: java.io.IOException: read failed: EINVAL (Invalid argument)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at libcore.io.IoBridge.read(IoBridge.java:481)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at java.io.FileInputStream.read(FileInputStream.java:252)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.os.ParcelFileDescriptor$AutoCloseInputStream.read(ParcelFileDescriptor.java:889)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at com.android.internal.os.BatteryStatsHelper.readFully(BatteryStatsHelper.java:733)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at com.android.internal.os.BatteryStatsHelper.getStats(BatteryStatsHelper.java:767)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at com.android.internal.os.BatteryStatsHelper.load(BatteryStatsHelper.java:755)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at com.android.internal.os.BatteryStatsHelper.getStats(BatteryStatsHelper.java:236)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at com.android.internal.os.BatteryStatsHelper.refreshStats(BatteryStatsHelper.java:323)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at com.android.internal.os.BatteryStatsHelper.refreshStats(BatteryStatsHelper.java:316)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at com.android.internal.os.BatteryStatsHelper.refreshStats(BatteryStatsHelper.java:309)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at com.android.settings.fuelgauge.PowerUsageBase.refreshStats(PowerUsageBase.java:124)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at com.android.settings.fuelgauge.PowerUsageSummary.refreshStats(PowerUsageSummary.java:281)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at com.android.settings.fuelgauge.PowerUsageSummary.onResume(PowerUsageSummary.java:100)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.app.Fragment.performResume(Fragment.java:2399)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1032)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1171)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1153)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.app.FragmentManagerImpl.dispatchResume(FragmentManager.java:2049)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.app.FragmentController.dispatchResume(FragmentController.java:198)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.app.Activity.performResume(Activity.java:6804)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3406)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3469)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2732)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.app.ActivityThread.-wrap12(ActivityThread.java)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1477)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.os.Handler.dispatchMessage(Handler.java:102)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.os.Looper.loop(Looper.java:154)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at android.app.ActivityThread.main(ActivityThread.java:6119)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at java.lang.reflect.Method.invoke(Native Method)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: Caused by: android.system.ErrnoException: read failed: EINVAL (Invalid argument)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at libcore.io.Posix.readBytes(Native Method)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at libcore.io.Posix.read(Posix.java:169)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at libcore.io.BlockGuardOs.read(BlockGuardOs.java:231)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	at libcore.io.IoBridge.read(IoBridge.java:471)
06-21 17:30:39.322   330   330 W BatteryStatsHelper: 	... 30 more
06-21 17:30:39.331   330   330 D AndroidRuntime: Shutting down VM
--------- beginning of crash
06-21 17:30:39.331   330   330 E AndroidRuntime: FATAL EXCEPTION: main
06-21 17:30:39.331   330   330 E AndroidRuntime: Process: com.android.settings, PID: 330
06-21 17:30:39.331   330   330 E AndroidRuntime: java.lang.RuntimeException: Unable to resume activity {com.android.settings/com.android.settings.Settings$PowerUsageSummaryActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'long com.android.internal.os.BatteryStatsImpl$StopwatchTimer.getTotalTimeLocked(long, int)' on a null object reference
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3429)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3469)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2732)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.app.ActivityThread.-wrap12(ActivityThread.java)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1477)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:102)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:154)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6119)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
06-21 17:30:39.331   330   330 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'long com.android.internal.os.BatteryStatsImpl$StopwatchTimer.getTotalTimeLocked(long, int)' on a null object reference
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at com.android.internal.os.BatteryStatsImpl.getPhoneOnTime(BatteryStatsImpl.java:4942)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at com.android.internal.os.BatteryStatsHelper.addPhoneUsage(BatteryStatsHelper.java:557)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at com.android.internal.os.BatteryStatsHelper.processMiscUsage(BatteryStatsHelper.java:677)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at com.android.internal.os.BatteryStatsHelper.refreshStats(BatteryStatsHelper.java:442)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at com.android.internal.os.BatteryStatsHelper.refreshStats(BatteryStatsHelper.java:316)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at com.android.internal.os.BatteryStatsHelper.refreshStats(BatteryStatsHelper.java:309)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at com.android.settings.fuelgauge.PowerUsageBase.refreshStats(PowerUsageBase.java:124)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at com.android.settings.fuelgauge.PowerUsageSummary.refreshStats(PowerUsageSummary.java:281)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at com.android.settings.fuelgauge.PowerUsageSummary.onResume(PowerUsageSummary.java:100)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.app.Fragment.performResume(Fragment.java:2399)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1032)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1171)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1153)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.app.FragmentManagerImpl.dispatchResume(FragmentManager.java:2049)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.app.FragmentController.dispatchResume(FragmentController.java:198)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.app.Activity.performResume(Activity.java:6804)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3406)
06-21 17:30:39.331   330   330 E AndroidRuntime: 	... 10 more
06-21 17:30:39.332   120   289 W ActivityManager:   Force finishing activity com.android.settings/.Settings$PowerUsageSummaryActivity
06-21 17:30:39.332   120   289 W ActivityManager:   Force finishing activity com.android.settings/.Settings
06-21 17:30:39.335   330   330 I Process : Sending signal. PID: 330 SIG: 9
```

ashmem and binder info:
```
$ modinfo ashmem_linux
filename:       /lib/modules/4.13.0-45-generic/updates/dkms/ashmem_linux.ko
license:        GPL
srcversion:     FE236C52294E4C6D2976CBE
depends:        
name:           ashmem_linux
vermagic:       4.13.0-45-generic SMP mod_unload 
```
```
$ modinfo binder_linux 
filename:       /lib/modules/4.13.0-45-generic/updates/dkms/binder_linux.ko
license:        GPL v2
srcversion:     B5AA82483799CFAE265C64B
depends:        
name:           binder_linux
vermagic:       4.13.0-45-generic SMP mod_unload 
parm:           debug_mask:uint
parm:           proc_no_lock:bool
parm:           num_devices:int
```

**Comment by raspimana** [comment]
When You Said “Setting”, Did You Mean “Settings”? Cc-raspimana

From The (Kinda) New Github Person, raspimana.

**Comment by raspimana** [comment]
When You Said “Setting”, Did You Mean “Settings”? Cc-raspimana
Fwd Cause I Deleted My Comment Om The Github Site.

From The (Kinda) New Github Person, raspimana.

Begin forwarded message:

> From: Aedan Reed <raspiman@icloud.com>
> Date: June 21, 2018 at 1:46:05 PM PDT
> To: anbox/anbox <reply@reply.github.com>
> Subject: Re: [anbox/anbox] Setting crashing due to battery stats (#767)
> 
> When You Said “Setting”, Did You Mean “Settings”? Cc-raspimana

**Comment by yihuaf** [comment]
Yes. Thanks for pointing that out :) I am referring to the Settings menu.

**Comment by yihuaf** [comment]
After digging around more, it seems that ashmem_linux is the cause to this. I manually compiled the ashmem module from this commit: `19110b824fc5110ac74494e294fb740a7a0ce38d` and it worked. 

I think the change to use kernel_read has something to do with it. More specifically, the only change between the TOT and the working version is this pull request: https://github.com/anbox/anbox/pull/504. I am not too familiar with the inner working of Ubuntu kernel, so please let me know if there is anything I can do or more information I can provide.

**Comment by raspimana** [comment]
You Welcome. I’m Always Here. :) (On My Phone.)

From The (Kinda) New Github Person, raspimana.

**Comment by ympavlov** [comment]
I had the same problem on Linux kernel 4.13 (ubuntu xenial). And I can't choose any application in Settings -> Applications — it crashes every time with the same error "NullPointerException: Attempt to invoke virtual method 'long com.android.internal.os.BatteryStatsImpl$StopwatchTimer.getTotalTimeLocked(long, int)' on a null object reference"
On Linux 4.15 it works.

**Comment by yihuaf** [comment]
The last time I investigated this, it has to do with the vfs_read and kernel read the kernel module was using. At some point, the vfs_read was deprecated, so it changed to kernel_read. @morphis If possible, can we have someone who familiar with that change take a look at this, since I am not the only one who can repro this? The fix/hack I did essentially is to use vfs_read until version 14 since that is the last kernel version where vfs_read is still exposed.
